text/0000-build-std.md

```diff
@@ -0,0 +1,2390 @@
+pre-built standard library from the sysroot.
+
+`-Zbuild-std` builds `std` by default. `test` is also built if tests are being
+run. Optionally, users can provide the list of crates to be built, though this
+was intended as an escape hatch to work around bugs - the arguments to the flag
+are inherently unstable since the names of crates comprising the standard
+library are not stable.
+
+Cargo has a hardcoded list of what dependencies need to be added for a given
+user-requested crate (i.e. `std` implies building `core`, `alloc`,
+`compiler_builtins`, etc.). It is common for users to manually specify the
+`panic_abort` crate.
+
+Originally, `-Zbuild-std` required that `--target` be provided
+([wg-cargo-std-aware#25]) to force Cargo to use different sysroots for the host
+and target , but this restriction was later resolved ([cargo#14317]).
+
+A second flag, [`-Zbuild-std-features`][build-std-features], was added in
+[cargo#8490] and allows overriding the default Cargo features of the standard
+library. Like the arguments to `-Zbuild-std`, this values accepted by this flag
+are inherently unstable as the library team has not committed to any of the
+standard library's Cargo features being stable. Features are enabled on the
+`sysroot` crate and propagate down through the crate graph of the standard
+library (e.g. `compiler-builtins-mem` is a feature in `sysroot`, `std`,
+`alloc`, and `core` until `compiler_builtins`).
+
+build-std gets the source of the standard library from the `rust-src` rustup
+component. This does not happen automatically and the user must ensure the
+component has been downloaded themselves. Only the standard library crates from
+the [rust-lang/rust] repository are included in the `rust-src` depdendency (i.e.
+none of the crates.io dependencies).
+
+When `-Zbuild-std` has been passed, Cargo creates a second workspace for the
+standard library based on the `Cargo.{toml,lock}` from the `rust-src` component.
+Originally this was a virtual workspace, prior to the standard library having a
+separate workspace from the compiler which could be used independently
+([rust#128534]/[cargo#14358]). This workspace is then resolved separately and
+the resolve is combined with the user's resolve to produce a dependency graph of
+things to build with the user's crates depending on the standard library's
+crates. Some additional work is done to deduplicate crates across the graph and
+then this crate graph is used to drive work (usually `rustc` invocations) as
+usual. This approach allows for build-time parallelism and sharing of crates
+between the two separate resolves but does involve a lot of `build-std`-specific
+logic in and around the resolver - an important and complicated part of Cargo.
+
+When first implemented, the multiple resolver approach was considered a
+short-term solution and that it was "quite messy and causes a lot of
+complications". While alternative solutions have been proposed
```

Comment:
I'm not sure if that is something I said, but it was never intended to be a short-term solution. Fairly early on we decided it was unavoidable.

It does result in that the standard library crates are otherwise hidden to the rest of cargo. For example `cargo tree`, `cargo metadata`, `cargo vendor` (since fixed) all act as-if the standard library doesn't exist. Each command needs special handling to support it. However, in many cases that might be what we want anyways.